### DISCUSSION TITLE
Nested stores with React Context

### DISCUSSION THREAD

**Discussion by sebastiaan-dev** [pmndrs/zustand#2226] [Q&A]
Hi,

I have a current setup where I have a global store defined using the slices pattern:

```ts
export const useStore = create<UISlice & GlobalStateSlice>()(
  immer((...a) => ({
    ...createUISlice(...a),
    ...createGlobalStateSlice(...a),
  }))
);
```

Now I have components which require initialisation data before being rendered. I do this by using React context as follows:

```ts
export const EntryProvider = ({ children, id }: EntryProviderProps) => {
  const [entry, updateEntry] = useStore((s) => [s.getEntry(id), s.updateEntry]);

  const storeRef = useRef<EntryStore>();
  if (!storeRef.current) {
    storeRef.current = createEntryStore({
      entry,
      updateEntry,
    });
  }

  return (
    <EntryContext.Provider value={storeRef.current}>
      {children}
    </EntryContext.Provider>
  );
};
```

Here id is a required initial value for a new store called `EntryStore`. Inside this store are functions which operate over 1 element coming from the global store. The `updateEntry` function is logic passed in from the global store. So that state changes can be reflected from the local `EntryStore` back in the global store when necessary. Now my issue with this is that I will need to inject every variable and function from the global state needed by the local store, which can be quite a lot. Is there a leaner way of injecting the global store into the local store?

I have not made this into one big store due to some variables being required on store creation of `EntryStore` (among other nested stores) so I do not have to make checks for state being possibly `undefined`.

A benefit of the current approach is that the `get()` space of  `EntryStore` is lean right now, but I'd rather have all functions, and state, from the global state available in `get()` from the `EntryStore`.

Best Regards

**Comment by dbritto-dev** [comment]
@sebastiaan-dev a simple way to do it what you want is by passing the global store and the entry. For instance: 

```ts
export const EntryProvider = ({ children, id }: EntryProviderProps) => {
  const [main, entry, updateEntry] = useStore((s) => [s, s.getEntry(id), s.updateEntry]);

  const storeRef = useRef<EntryStore>();
  if (!storeRef.current) {
    storeRef.current = createEntryStore({
      entry,
      updateEntry,
    });
  }

  return (
    <EntryContext.Provider value={[main, entry]}>
      {children}
    </EntryContext.Provider>
  );
};
```

**Reply by sebastiaan-dev** [reply]
It seems a part of the reply is missing?

**Reply by dbritto-dev** [reply]
@sebastiaan-dev updated

**Reply by sebastiaan-dev** [reply]
Thank you for the quick response, I am looking for something different I think. I want to be able to do the following:

```ts
export const createEntryStore = (
  initProps: InitEntryProps & InitEntryState
) => {
  return createStore<EntryState & UISlice & GlobalStateSlice>()(
    immer((set, get, store) => ({
      ...initProps,
      
      // local api
      doSomething: () => {
        ...
        // access global store function with global state context
        return get().updateEntry(...);
      },
    }))
  );
};
```

How would you go about implementing this?

I can destructure the global slice in the `EntryStore` but then the state is not shared:

```ts
export const createEntryStore = (
  initProps: InitEntryProps & InitEntryState
) => {
  return createStore<EntryState & UISlice & GlobalStateSlice>()(
    immer((set, get, store) => ({
      ...globalStateSlice([set, get, store]),
      ...initProps,
      
      // local api
      doSomething: () => {
        ...
        // access global store function with global state context
        return get().updateEntry(...);
      },
    }))
  );
};
```

**Reply by dbritto-dev** [reply]
@sebastiaan-dev yeah, you can move everything under a namespace called a create slices

**Reply by sebastiaan-dev** [reply]
@dbritto-dev I am still struggling with this, if I do as below:

```ts
export const EntryProvider = ({ children, id }: EntryProviderProps) => {
  const [entry, updateEntry] = useStore((s) => [s.getEntry(id), s.updateEntry]);
  // Variable containing the entire global store
  const globalStore = useStore();

  const storeRef = useRef<EntryStore>();
  if (!storeRef.current) {
                                    // I inject the global store into the EntryStore
    storeRef.current = createEntryStore(globalStore, {
      entry,
      updateEntry,
    });
  }

  return (
    <EntryContext.Provider value={storeRef.current}>
      {children}
    </EntryContext.Provider>
  );
};
```

And then use it as follows in `EntryStore`:

```ts
export const createEntryStore = (
  globalStore: GlobalStateSlice
  initProps: InitEntryProps & InitEntryState
) => {
  return createStore<EntryState & UISlice & GlobalStateSlice>()(
    immer((set, get, store) => ({
      ...globalStore,
      ...initProps,
      
      // local api
      createSomething: () => {
        ...
        set((s) => {
          // This function comes from the global store slice
          s.createSomethingInGlobal(...);
         })
      },
      getGlobalSomethings: () => {
        ...
        // This variable is not stored in `EntryStore` but comes from the global store
        return get().globalSomethings;
      },
    }))
  );
};
```

The issue here is that if I `console.log(globalSomethings)` in `EntryStore` it misses the update from `createSomething`. When I log the`globalSomethings` inside the `createSomethingInGlobal` function after it has been created it does show up in the state. So the global and local state are out of sync. After rerendering the component (by leaving the page etc) the local state does show the updated `globalSomethings`. Any tips?

**Reply by sebastiaan-dev** [reply]
I can do something as below, but that might defeat the purpose of Zustand entirely? As in, cause unnecessary renders.

```ts
storeRef.current = useMemo(() => {
    return createEntryStore(globalStore, { entry, updateEntry });
  }, [globalStore]);
```

**Reply by dbritto-dev** [reply]
@sebastiaan-dev you would mind creating a minimal repro?
